src/common/upload.ts

```diff
@@ -118,36 +125,80 @@ export abstract class TurboAuthenticatedBaseUploadService
-    this.logger.debug('Uploading signed data item...');
-    // TODO: add p-limit constraint or replace with separate upload class
+    let retries = 0;
+    const maxRetries = this.retryConfig.retries ?? 3;
+    const retryDelay =
+      this.retryConfig.retryDelay ??
+      ((retryNumber: number) => retryNumber * 1000);
+    let lastError = 'Unknown Error'; // Store the last error for throwing
+    let lastStatusCode: number | undefined; // Store the last status code for throwing
+
+    while (retries < maxRetries) {
+      if (signal?.aborted) {
+        throw new CanceledError();
+      }
 
-    const headers = {
-      'content-type': 'application/octet-stream',
-      'content-length': `${dataItemSizeFactory()}`,
-    };
-    if (dataItemOpts !== undefined && dataItemOpts.paidBy !== undefined) {
-      const paidBy = Array.isArray(dataItemOpts.paidBy)
-        ? dataItemOpts.paidBy
-        : [dataItemOpts.paidBy];
+      const { dataItemStreamFactory, dataItemSizeFactory } =
+        await this.signer.signDataItem({
+          fileStreamFactory,
+          fileSizeFactory,
+          dataItemOpts,
+        });
+
+      try {
+        this.logger.debug('Uploading signed data item...');
+        // TODO: add p-limit constraint or replace with separate upload class
+
+        const headers = {
+          'content-type': 'application/octet-stream',
+          'content-length': `${dataItemSizeFactory()}`,
+        };
+        if (dataItemOpts !== undefined && dataItemOpts.paidBy !== undefined) {
+          const paidBy = Array.isArray(dataItemOpts.paidBy)
+            ? dataItemOpts.paidBy
+            : [dataItemOpts.paidBy];
+
+          if (dataItemOpts.paidBy.length > 0) {
+            headers['x-paid-by'] = paidBy;
+          }
+        }
+        const data = await this.httpService.post<TurboUploadDataItemResponse>({
+          endpoint: `/tx/${this.token}`,
+          signal,
+          data: dataItemStreamFactory(),
+          headers,
+        });
+        return data;
+      } catch (error) {
+        if (error instanceof FailedRequestError) {
+          throw error; // Rethrow if it's already a failed request error from HTTP service
+        }
 
-      if (dataItemOpts.paidBy.length > 0) {
-        headers['x-paid-by'] = paidBy;
+        // Store the last encountered error and status for re-throwing after retries
+        if (error instanceof AxiosError) {
+          lastStatusCode = error.response?.status;
+          lastError = error.code ?? error.message;
+        } else {
+          lastError =
+            error instanceof Error && error.message !== undefined
+              ? error.message
+              : `${error}`; // Stringify the whole error if it's not a known Error instance
+        }
+        this.logger.debug(
+          `Upload failed, attempt ${retries + 1}/${maxRetries}`,
+          { message: lastError },
+          error,
+        );
+        retries++;
+        await sleep(retryDelay(retries, error));
```

Comment:
Is `signal` eventable? If so you can wrap the abort event with a promise and then race it against sleep so that aborts will be performant. you'd have to check the result of the race to see which promise won (e.g. check if the abort handle is signaled)

Reply:
:notbad:

Reply:
c4c2617